### DISCUSSION TITLE
Milvus MySQL Access Optimization

### DISCUSSION THREAD

**Discussion by lsgrep** [milvus-io/milvus#13887] [Q&A and General discussion]
Hi, we are happily using Milvus 1.1 in production, and thank you for open-sourcing the  Milvus for the community.

However, in business-critical conditions, we require extremely low latency. I am reading through the source code for `search` and there are a couple of places that the engine access `mysql` for metadata. Is there a way to optimize this? 

Please correct me if I am wrong, I don't have much C++ experience.

1. In `GRPCRequestHandler.cpp`
Search begins with empty vector `file_ids`
<img width="1030" alt="Screen Shot 2021-12-21 at 5 28 48 PM" src="https://user-images.githubusercontent.com/3893940/146905826-4581651e-e5f3-4fc6-a3c9-d3fd46ea6f5c.png">

2. In `RequestHandler.cpp` , engine creates a new request & put it a queue then executes it asynchronously.
<img width="1095" alt="Screen Shot 2021-12-21 at 5 34 58 PM" src="https://user-images.githubusercontent.com/3893940/146906710-24b27c8d-d587-4e37-b53d-ae5465cb5f99.png">

3. In `SearchRequest.cpp` there is a `DescribeCollection` which require DB access for the first time.
<img width="1058" alt="Screen Shot 2021-12-21 at 5 40 48 PM" src="https://user-images.githubusercontent.com/3893940/146907622-f4933ce9-8cde-4adc-9f2c-c1df5e5e1438.png">

Since the `file_id_list_` is empty, there will be another 2 DB access for the second time.(`FilesToSearch`, `ShowPartitions`)
<img width="1035" alt="Screen Shot 2021-12-21 at 5 42 16 PM" src="https://user-images.githubusercontent.com/3893940/146907864-118c5e32-f9bd-4cdc-8c04-05ae46ec3c76.png">
<img width="830" alt="Screen Shot 2021-12-21 at 5 56 09 PM" src="https://user-images.githubusercontent.com/3893940/146909993-1259ec88-a047-4640-9b17-2ecd40a9ffcb.png">

So there will 3 DB queries for each search?  And the first `DescribeCollection` is not necessary since the subsequent queries will fail if the `collection` does not exist?

**Comment by yhmo** [comment]
To reduce access to MySQL, we need cache machinery for metadata (but we have stopped developing new features for 1.x) . So far as I know, some users develop such kind of metadata cache based on v1.1 as their private edition, and it did increase search performance.

The DescribeCollection() in the SearchRequest is to validate search parameters such as: the input target vector's dimension is equal to collection's dimension? the search parameters "nprobe" "topk" is in range?
